File: app/src/main/java/br/com/jonathanzanella/myexpenses/bill/BillAdapterPresenter.java
package br.com.jonathanzanella.myexpenses.bill;

import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class BillAdapterPresenter {
	private BillRepository repository;

	private List<Bill> bills;
	private List<Bill> billsFiltered;

	BillAdapterPresenter(BillRepository repository) {
		this.repository = repository;
		loadBills();
	}

	private void loadBills() {
		bills = repository.userBills();
		billsFiltered = bills;
	}

	List<Bill> getBills(boolean invalidateCache) {
		if(invalidateCache)
			loadBills();
		return Collections.unmodifiableList(billsFiltered);
	}

	public void filter(String filter) {
		if(filter == null || filter.compareTo("") == 0) {
			billsFiltered = bills;
			return;
		}

		billsFiltered = new ArrayList<>();
		for (Bill bill : bills) {
			if(StringUtils.containsIgnoreCase(bill.getName(), filter))
				billsFiltered.add(bill);
		}
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/bill/BillAdapterPresenter.kt
package br.com.jonathanzanella.myexpenses.bill

import org.apache.commons.lang3.StringUtils
import java.util.*

class BillAdapterPresenter(private val repository: BillRepository) {
    private var bills: MutableList<Bill>? = null
    private var billsFiltered: MutableList<Bill>? = null

    init {
        loadBills()
    }

    private fun loadBills() {
        bills = repository.all() as MutableList<Bill>
        billsFiltered = bills
    }

    fun getBill(position: Int): Bill {
        return billsFiltered!![position]
    }

    fun getBills(invalidateCache: Boolean): List<Bill> {
        if (invalidateCache)
            loadBills()
        return Collections.unmodifiableList(billsFiltered!!)
    }

    val billsCount: Int
        get() = billsFiltered!!.size

    fun filter(filter: String) {
        if (filter.compareTo("") == 0) {
            billsFiltered = bills
            return
        }

        billsFiltered = ArrayList<Bill>()
        bills!!
                .filter { StringUtils.containsIgnoreCase(it.name, filter) }
                .forEach { billsFiltered!!.add(it) }
    }
}
